shared/src/pool_fetching.rs

```diff
@@ -299,57 +315,93 @@ impl PoolFetching for PoolFetcher {
+                    .block(at_block.into())
+                    .batch_call(&mut batch);
+                let token1_balance = token1
+                    .balance_of(pair_address)
+                    .block(at_block.into())
+                    .batch_call(&mut batch);
+
+                async move {
+                    // Clippy is wrong about this being eval order depndendent.
+                    #[allow(clippy::eval_order_dependence)]
+                    FetchedPool {
+                        pair,
+                        reserves: reserves.await,
+                        token0_balance: token0_balance.await,
+                        token1_balance: token1_balance.await,
+                    }
+                }
             })
             .collect::<Vec<_>>();
-
         batch.execute_all(MAX_BATCH_SIZE).await;
 
-        let mut results = Vec::with_capacity(futures.len());
-        for (pair, get_reserves, token0_balance, token1_balance) in futures {
-            let reserves = get_reserves.await;
-            let token0_balance = token0_balance.await;
-            let token1_balance = token1_balance.await;
-            if let (Ok(reserves), Ok(token0_balance), Ok(token1_balance)) =
-                (reserves, token0_balance, token1_balance)
-            {
-                // Some ERC20s (e.g. AMPL) have an elastic supply and can thus reduce the balance of their owners without any transfer or other interaction ("rebase").
-                // Such behavior can implicitly change the *k* in the pool's constant product formula. E.g. a pool with 10 USDC and 10 AMPL has k = 100. After a negative
-                // rebase the pool's AMPL balance may reduce to 9, thus k should be implicitly updated to 90 (figuratively speaking the pool is undercollateralized).
-                // Uniswap pools however only update their reserves upon swaps. Such an "out of sync" pool has numerical issues when computing the right clearing price.
-                // Note, that a positive rebase is not problematic as k would increase in this case giving the pool excess in the elastic token (an arbitrageur could
-                // benefit by withdrawing the excess from the pool without selling anything).
-                // We therefore exclude all pools where the pool's token balance of either token in the pair is less than the cached reserve.
-                if U256::from(reserves.0) <= token0_balance
-                    && U256::from(reserves.1) <= token1_balance
-                {
-                    results.push(Pool::uniswap(pair, (reserves.0, reserves.1)));
-                }
-            }
+        let mut results = Vec::new();
+        for future in futures {
+            results.push(future.await);
+        }
+        handle_results(results)
+    }
+}
+
+struct FetchedPool {
+    pair: TokenPair,
+    reserves: Result<(u128, u128, u32), MethodError>,
+    token0_balance: Result<U256, MethodError>,
+    token1_balance: Result<U256, MethodError>,
+}
```

Comment:
I wanted to ask here why we keep track of both `reserves` and `token_i_balance`. Shouldn't the reserves already contain these balances? Also, may I assume that the `u32` in the reserves tuple is the `block_last_updated`? Could we benefit from making use of this instead of all the fancy block logic?

Reply:
Oh I see in the comment that we 

```
// Fetch ERC20 token balances of the pools to sanity check with reserves
```

However I would think that we could perform the sanity check upon pool fetching and then forget about these values.

Reply:
This PR didn't change that logic just refactored it so maybe let's discuss it in an issue or on slack so it's not hidden here.